File: sdk/csharp/Rin.Client/RinException.cs
namespace Rin.Client;

public class RinException : Exception
{
    public RinException(string code, string message, Exception? innerException = null)
        : base(SafeText(message, 500, "Rin request failed"), innerException)
    {
        Code = SafeText(code, 96, "rin_error");
    }

    public string Code { get; }

    internal static string SafeText(string? value, int maximum, string fallback = "")
    {
        var cleaned = string.Join(" ", (value ?? string.Empty)
            .Replace("\0", string.Empty, StringComparison.Ordinal)
            .Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
        return cleaned.Length == 0 ? fallback : cleaned[..Math.Min(cleaned.Length, maximum)];
    }
}

public sealed class RinConfigurationException : RinException
{
    public RinConfigurationException(string code, string message, Exception? innerException = null)
        : base(code, message, innerException) { }
}

public sealed class RinTransportException : RinException
{
    public RinTransportException(string code, string message, Exception? innerException = null)
        : base(code, message, innerException) { }
}

public sealed class RinProtocolException : RinException
{
    public RinProtocolException(string code, string message, Exception? innerException = null)
        : base(code, message, innerException) { }
}

public sealed class RinApiException : RinException
{
    public RinApiException(string code, string message, int status = 0, string? field = null)
        : base(code, message)
    {
        Status = status;
        Field = SafeText(field, 160);
    }

    public int Status { get; }

    public string Field { get; }
}

File: sdk/java/src/main/java/io/github/sunrioa/rin/RinException.java
package io.github.sunrioa.rin;

public class RinException extends RuntimeException {
    private static final long serialVersionUID = 1L;

    private final String code;

    public RinException(String code, String message) {
        this(code, message, null);
    }

    public RinException(String code, String message, Throwable cause) {
        super(safeText(message, 500, "Rin request failed"), cause);
        this.code = safeText(code, 96, "rin_error");
    }

    public String code() {
        return code;
    }

    static String safeText(Object value, int maximum, String fallback) {
        String cleaned = String.valueOf(value == null ? "" : value)
                .replace('\0', ' ')
                .trim()
                .replaceAll("\\s+", " ");
        if (cleaned.isEmpty()) return fallback;
        return cleaned.substring(0, Math.min(cleaned.length(), maximum));
    }
}
